ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java

```diff
@@ -88,64 +78,52 @@ public static void updateEth1Data(BeaconState state) throws EpochProcessingExcep
           ValidatorsUtil.get_active_validator_indices(
               state.getValidator_registry(), previous_epoch);
-      UnsignedLong current_total_balance =
+      long current_total_balance =
           BeaconStateUtil.get_total_balance(state, current_active_validators);
-      UnsignedLong previous_total_balance =
+      long previous_total_balance =
           BeaconStateUtil.get_total_balance(state, previous_active_validators);
 
       // Update justification bitfield
-      UnsignedLong new_justified_epoch = state.getJustified_epoch();
-      UnsignedLong justification_bitfield = state.getJustification_bitfield();
-      justification_bitfield = BitwiseOps.leftShift(justification_bitfield, 1);
-
-      if (AttestationUtil.get_previous_epoch_boundary_attesting_balance(state)
-              .times(UnsignedLong.valueOf(3))
-              .compareTo(previous_total_balance.times(UnsignedLong.valueOf(2)))
-          >= 0) {
-        justification_bitfield = BitwiseOps.or(justification_bitfield, UnsignedLong.valueOf(2));
+      long new_justified_epoch = state.getJustified_epoch();
+      long justification_bitfield = state.getJustification_bitfield();
+      justification_bitfield = justification_bitfield << 1;
+
+      if (AttestationUtil.get_previous_epoch_boundary_attesting_balance(state) * 3
+          >= previous_total_balance * 2) {
+        justification_bitfield = justification_bitfield | 2;
         new_justified_epoch = previous_epoch;
       }
-      if (AttestationUtil.get_current_epoch_boundary_attesting_balance(state)
-              .times(UnsignedLong.valueOf(3))
-              .compareTo(current_total_balance.times(UnsignedLong.valueOf(2)))
-          >= 0) {
-        justification_bitfield = BitwiseOps.or(justification_bitfield, UnsignedLong.ONE);
+      if (AttestationUtil.get_current_epoch_boundary_attesting_balance(state) * 3
+          >= current_total_balance * 2) {
+        justification_bitfield = justification_bitfield | 1;
         new_justified_epoch = current_epoch;
       }
 
       state.setJustification_bitfield(justification_bitfield);
 
       // Update last finalized epoch if possible
-      UnsignedLong decimal4 = UnsignedLong.valueOf(4);
-      UnsignedLong decimal8 = UnsignedLong.valueOf(8);
-      UnsignedLong binary11 = UnsignedLong.valueOf(3);
-      UnsignedLong binary111 = UnsignedLong.valueOf(7);
-      UnsignedLong previous_justified_epoch = state.getPrevious_justified_epoch();
-      UnsignedLong justified_epoch = state.getJustified_epoch();
-
-      if (BitwiseOps.rightShift(justification_bitfield, 1).mod(decimal8).equals(binary111)
-          && previous_justified_epoch.equals(previous_epoch.minus(UnsignedLong.valueOf(2)))) {
+      long previous_justified_epoch = state.getPrevious_justified_epoch();
+      long justified_epoch = state.getJustified_epoch();
+
+      if (justification_bitfield >> 1 % 8 == 7 && previous_justified_epoch == previous_epoch - 2) {
```

Comment:
This should probably be an unsigned right shift instead of a signed right shift.